src/types/profile.js

```diff
@@ -20,9 +20,36 @@ export type resourceTypeEnum = number;
 export type ThreadIndex = number;
 
 /**
- * The stack table is the minimal representation of a call stack. Each stack entry
- * consists of the frame at the top of the stack, and the prefix for the stack that
- * came before it. Stacks can be shared between samples.
+ * The stack table stores the tree of stack nodes of a thread.
+ * The shape of the tree is encoded in the prefix column: Root stack nodes have
+ * null as their prefix, and every non-root stack has the stack index of its
+ * "caller" / "parent" as its prefix.
+ * Every stack node also has a frame and a category.
+ * A "call stack" is a list of frames. Every stack index in the stack table
+ * represents such a call stack; the "list of frames" is obtained by walking
+ * the path in the tree from the root to the given stack node.
+ *
+ * Stacks are used in the thread's samples; each sample refers to a stack index.
+ * Stacks can be shared between samples.
+ *
+ * With this representation, every sample only needs to store a single integer
+ * to identify the sample's stack.
+ * We take advantage of the fact that many call stacks in the profile have a
+ * shared prefix; storing these stacks as a tree saves a lot of space compared
+ * to storing them as actual lists of frames.
+ *
+ * The category of a stack node is always non-null and is derived from a stack's
+ * frame and its prefix. Frames can have null categories, stacks cannot. If a
+ * stack's frame has a null category, the stack inherits the category of its
+ * prefix stack. Root stacks whose frame has a null stack have their category
+ * set to the "default category". (The default category is currently defined as
+ * the category in the profile's category list whose color is "grey", and such
+ * a category is required to be present.)
+ *
+ * You could argue that the stack table's category column is derived data and as
+ * such doesn't need to be stored in the profile itself. This is true, but
+ * storing this information in the stack table makes it a lot easier to carry
+ * it through various transforms that we apply to threads.
```

Comment:
The alternative is recomputing it from the transformed frame information. This would allow that we don't really mind about it, but we couldn't likely manage conflicts as easily (which I'm still not sure we should).  

I'll trust you here as you thought about this more than I did.

Reply:
I'll add a sentence to this comment that gives an example of where recomputing it from the transformed frame information would not be good enough.

Reply:
I'm adding the following sentence to the comment:

```
 * For example, here's a case where a stack's category is not recoverable from
 * any other information in the transformed thread:
 * In partial the call path
 *   someJSFunction [JS] -> Node.insertBefore [DOM] -> nsAttrAndChildArray::InsertChildAt,
 * the stack node for nsAttrAndChildArray::InsertChildAt should inherit the
 * category DOM from its "Node.insertBefore" prefix stack. And it should keep
 * the DOM category even if you apply the "Merge node into calling function"
 * transform to Node.insertBefore. This transform removes the stack node
 * "Node.insertBefore" from the stackTable, so the information about the DOM
 * category would be lost if it wasn't inherited into the
 * nsAttrAndChildArray::InsertChildAt stack before transforms are applied.
```